pkg/solana/logpoller/loader.go

```diff
@@ -1,477 +1,163 @@
+	workers *worker.Group
 }
 
 func NewEncodedLogCollector(
 	client RPCClient,
-	parser ProgramEventProcessor,
-	lggr logger.Logger,
+	lggr logger.SugaredLogger,
 ) *EncodedLogCollector {
 	c := &EncodedLogCollector{
-		client:       client,
-		unordered:    newUnorderedParser(parser),
-		chSlot:       make(chan uint64),
-		chBlock:      make(chan uint64, 1),
-		chJobs:       make(chan Job, 1),
-		lggr:         lggr,
-		rpcTimeLimit: 1 * time.Second,
+		client: client,
+		lggr:   lggr,
 	}
 
 	c.Service, c.engine = services.Config{
 		Name: "EncodedLogCollector",
 		NewSubServices: func(lggr logger.Logger) []services.Service {
-			c.workers = NewWorkerGroup(DefaultWorkerCount, lggr)
-			c.ordered = newOrderedParser(parser, lggr)
+			c.workers = worker.NewGroup(worker.DefaultWorkerCount, logger.Sugared(lggr))
 
-			return []services.Service{c.workers, c.ordered}
+			return []services.Service{c.workers}
 		},
-		Start: c.start,
-		Close: c.close,
 	}.NewServiceEngine(lggr)
 
 	return c
 }
 
-func (c *EncodedLogCollector) BackfillForAddress(ctx context.Context, address string, fromSlot uint64) error {
-	pubKey, err := solana.PublicKeyFromBase58(address)
-	if err != nil {
-		return err
+func (c *EncodedLogCollector) getSlotsToFetch(ctx context.Context, addresses []PublicKey, fromSlot, toSlot uint64) ([]uint64, error) {
+	// identify slots to fetch
+	slotsForAddressJobs := make([]*getSlotsForAddressJob, len(addresses))
+	slotsToFetch := make(map[uint64]struct{}, toSlot-fromSlot)
+	var slotsToFetchMu sync.Mutex
+	storeSlot := func(slot uint64) {
+		slotsToFetchMu.Lock()
+		slotsToFetch[slot] = struct{}{}
+		slotsToFetchMu.Unlock()
 	}
-
-	var (
-		lowestSlotRead uint64
-		lowestSlotSig  solana.Signature
-	)
-
-	for lowestSlotRead > fromSlot || lowestSlotRead == 0 {
-		opts := rpc.GetSignaturesForAddressOpts{
-			Commitment:     rpc.CommitmentFinalized,
-			MinContextSlot: &fromSlot,
-		}
-
-		if lowestSlotRead > 0 {
-			opts.Before = lowestSlotSig
-		}
-
-		sigs, err := c.client.GetSignaturesForAddressWithOpts(ctx, pubKey, &opts)
+	for i, address := range addresses {
+		slotsForAddressJobs[i] = newGetSlotsForAddress(c.client, c.workers, storeSlot, address, fromSlot, toSlot)
+		err := c.workers.Do(ctx, slotsForAddressJobs[i])
 		if err != nil {
-			return err
-		}
-
-		if len(sigs) == 0 {
-			break
-		}
-
-		// signatures ordered from newest to oldest, defined in the Solana RPC docs
-		for _, sig := range sigs {
-			lowestSlotSig = sig.Signature
-
-			if sig.Slot >= lowestSlotRead && lowestSlotRead != 0 {
-				continue
-			}
-
-			lowestSlotRead = sig.Slot
-
-			if err := c.workers.Do(ctx, &getTransactionsFromBlockJob{
-				slotNumber: sig.Slot,
-				client:     c.client,
-				parser:     c.unordered,
-				chJobs:     c.chJobs,
-			}); err != nil {
-				return err
-			}
+			return nil, fmt.Errorf("could not shedule job to fetch slots for address: %w", err)
 		}
 	}
 
-	return nil
-}
-
-func (c *EncodedLogCollector) start(_ context.Context) error {
```

Comment:
I know there was discussion on changing the strategy for reading block data, but I see a lot of stuff removed and no proof that the change was an improvement.

Reply:
The main goal of this PR is to deliver a functioning LogPoller. We can alwaysys reintroduce forward-looking processing as a followup.

Reply:
I'm not sure delivering a functioning LogPoller required removing functionality though. Yes, we can reintroduce forward-looking processing, but I would expect that would be more difficult than leaving it in.

Reply:
If you insist, I can roll back removal, but I'm not sure that leaving unused code is a good idea.

Reply:
Well I'm not going to stonewall this, but I just think the approach of removing something for the sake of efficiency for one product and sacrificing future capability is generally bad practice.